consensus/istanbul/config.go

```diff
@@ -16,29 +16,99 @@
 
 package istanbul
 
-import "math/big"
+import (
+	"math/big"
 
-type ProposerPolicy uint64
+	"github.com/naoina/toml"
+)
+
+type ProposerPolicyId uint64
 
 const (
-	RoundRobin ProposerPolicy = iota
+	RoundRobin ProposerPolicyId = iota
 	Sticky
 )
 
+// ProposerPolicy represents the Validator Proposer Policy
+type ProposerPolicy struct {
+	Id       ProposerPolicyId    // Could be RoundRobin or Sticky
+	Registry []ValidatorSet      // Holds the ValidatorSet for a given block height
+	By       validatorSortByFunc // func that defines how the ValidatorSet should be sorted
+}
+
+// NewRoundRobinProposerPolicy returns a RoundRobin ProposerPolicy with ValidatorSortByStringFunc as default sort function
+func NewRoundRobinProposerPolicy() *ProposerPolicy {
+	return NewProposerPolicy(RoundRobin)
+}
+
+// NewStickyProposerPolicy return a Sticky ProposerPolicy with ValidatorSortByStringFunc as default sort function
+func NewStickyProposerPolicy() *ProposerPolicy {
+	return NewProposerPolicy(Sticky)
+}
+
+func NewProposerPolicy(id ProposerPolicyId) *ProposerPolicy {
+	return &ProposerPolicy{Id: id, By: ValidatorSortByStringFunc}
+}
+
+type proposerPolicyToml struct {
+	Id ProposerPolicyId
+}
+
+func (p *ProposerPolicy) MarshalTOML() ([]byte, error) {
+	pp := &proposerPolicyToml{Id: p.Id}
+	return toml.Marshal(pp)
+}
+
+func (p *ProposerPolicy) UnmarshalTOML(input []byte) error {
+	var pp proposerPolicyToml
+	err := toml.Unmarshal(input, &pp)
+	if err != nil {
+		return err
+	}
+	p.Id = pp.Id
+	p.By = ValidatorSortByStringFunc
+	return nil
+}
+
+// Use sets the validatorSortByFunc for the given ProposerPolicy and sorts the validatorSets according to it
+func (p *ProposerPolicy) Use(v validatorSortByFunc) {
```

Comment:
might be worth adding some unit tests for this because it requires pointers to always be passed around. A unit test will protect against a potential future change by an unsuspecting dev passing vals around

Reply:
Added a test for ProposerPolicy